hilti/toolchain/src/compiler/visitors/normalizer.cc

```diff
@@ -0,0 +1,236 @@
+        if ( n.init() && ! n.condition() ) {
+            auto cond = expression::UnresolvedID(n.init()->id());
+            logChange(p.node, cond);
+            p.node.as<statement::If>().setCondition(std::move(cond));
+            modified = true;
+        }
+    }
+
+    void operator()(const statement::Switch& s, position_t p) { p.node.as<statement::Switch>().preprocessCases(); }
+
+    void operator()(const type::Library& t, position_t p) {
+        auto& type = p.node.as<Type>();
+
+        if ( ! type.cxxID() )
+            // Make it equal to types with the same C++ representation.
+            type.setCxxID(ID(t.cxxName()));
+    }
+
+    void operator()(const type::Struct& t, position_t p) { type::Struct::setSelf(&p.node); }
+};
+
+} // anonymous namespace
+
+// Visitor to unset all canonical IDs in preparation for their recalculation.
+struct VisitorClearCanonicalIDs : public visitor::PreOrder<void, VisitorClearCanonicalIDs> {
+    result_t operator()(const Declaration& d, position_t p) { p.node.as<Declaration>().setCanonicalID(ID()); };
+};
+
+// Visitor computing canonical IDs.
+struct VisitorComputeCanonicalIDs : public visitor::PreOrder<ID, VisitorComputeCanonicalIDs> {
+    ID parent_id;
+    ID module_id;
+    Scope* module_scope = nullptr;
+
+    result_t operator()(const Module& m, position_t p) {
+        module_id = m.id();
+        module_scope = p.node.scope().get();
+        return m.id();
+    }
+
+    result_t operator()(const Declaration& d, position_t p) {
+        ID id;
+
+        // A couple of special-cases for top-level declarations.
+        if ( parent_id.length() == 1 ) {
+            // 1. If the ID is qualified with the current module, the ID is
+            // fine as it is.
+            if ( d.id().sub(0) == module_id )
+                id = d.id();
+
+            // 2. If the ID refers to something inside an imported module, we
+            // likewise use the ID as it is.
+            else if ( auto x = module_scope->lookup(d.id().sub(0)); x && x->node->isA<declaration::ImportedModule>() )
+                id = d.id();
+        }
+
+        if ( auto x = d.tryAs<declaration::ImportedModule>() )
+            // Use the namespace to the imported module as our ID.
+            id = x->id();
+
+        if ( ! id )
+            // By default, prefix the ID with the current parent.
+            id = ID(parent_id, d.id());
+
+        // Record the ID. If multiple paths lead to an ID, take the shortest.
+        if ( ! d.canonicalID() || id.length() < d.canonicalID().length() )
+            p.node.as<Declaration>().setCanonicalID(id);
+
+        assert(d.canonicalID());
+        return d.canonicalID();
+    }
+};
+
+// Visitor double-checking that all declarations have their canonical IDs set.
+struct VisitorCheckCanonicalIDs : public visitor::PreOrder<void, VisitorCheckCanonicalIDs> {
+    result_t operator()(const Declaration& d, position_t p) { assert(d.canonicalID()); };
```

Comment:
If I also add the following check here, e.g. the test `spicy.types.unit.hooks-priorities` fails with an assertion error:

```.cpp
    result_t operator()(const expression::ResolvedID& x, position_t p) { assert(x.declaration().canonicalID()); }
```

Reply:
Ok, pushed a fix. Not seeing any other missing canon IDs right now other than then struct constructor (which, as said, is expected).

Reply:
As of 1091d61 this became much more unstable and I see a lot more tests fail now with optimizations. I checked that even without my patches already your branch leaves a number of canon IDs unset which probably should be set, e.g., just for the HILTI tests:

```
$ for f in $(find tests/hilti -name '*.hlt'); do 
    build/bin/hiltic -p $f -o /dev/null -D ast-final 2>&1 \
        | grep -q 'canon-id: not set' && echo "UNSET IN $f"
done
```

I set hits from tests not involving structs here, e.g., `tests/hilti/expressions/list-comprehension.hlt`.

Reply:
Likewise fixed now.